include/irr/asset/filters/CSummedAreaTableImageFilter.h

```diff
@@ -0,0 +1,382 @@
+			};
+
+			allocateCache(state->rowSumCache, outExtent.X * texelByteSize);
+			allocateCache(state->columnSumCache, outExtent.Y * texelByteSize);
+
+			auto perOutputRegion = [state]( const CMatchedSizeInOutImageFilterCommon::CommonExecuteData& commonExecuteData, CBasicImageFilterCommon::clip_region_functor_t& clip) -> bool
+			{
+				const auto blockDims = asset::getBlockDimensions(inFormat);
+
+				auto sum = [&commonExecuteData, &blockDims](uint32_t readBlockArrayOffset, core::vectorSIMDu32 readBlockPos) -> void
+				{
+					auto localOutPos = readBlockPos * blockDims + commonExecuteData.offsetDifference; 
+					auto outPtrTexelPosition = commonExecuteData.outData + commonExecuteData.oit->getByteOffset(localOutPos, commonExecuteData.outByteStrides);
+					auto inPtrTexelPosition = commonExecuteData.inData + readBlockArrayOffset;
+
+					constexpr auto MAX_PLANES = 4;
+					const void* srcInPix[MAX_PLANES] = { inPtrTexelPosition, nullptr, nullptr, nullptr };
+					uint8_t* dstPix = commonExecuteData.outData + commonExecuteData.oit->getByteOffset(localOutPos, commonExecuteData.outByteStrides);
+
+					// with caching we can just substract one value (looking at left or down) to get appropriate sum without summing everything many times per texel 
+
+					// fill cache with max values
+					auto fillCacheIfEnteredMaxValues = [&]()
+					{
+						const auto width = commonExecuteData.oit->imageExtent.width;
+						const auto height = commonExecuteData.oit->imageExtent.height;
+						core::vector4df_SIMD channels; // tmp, TODO
+
+						//convertColor<inFormat, outFormat>(srcPix, dstPix, blockX, blockY); nah		
+
+						if (localOutPos.X == width - 1 && localOutPos.Y == height - 1)
+						{
+							double finalMaxSumForTexel = {};
+
+							// encode and decode it properly
+							for (size_t yPos = 0; yPos < height; ++yPos)	// compute max sum values in rows for y texel sum values
+							{
+								// use finalMaxSumForTexel and pointer arythmetic to calculate final sum for a texel, towards LEFT
+								// gain max sum for X values per each y
+							}
+
+							for (size_t xPos = 0; xPos < width; ++xPos)
+							{
+								// use finalMaxSumForTexel and pointer arythmetic to calculate final sum for a texel, towards DOWN
+								// gain max sum for Y values per each X
+							}
+
+							if (!ExclusiveMode)
+							{
+								// add to finalMaxSumForTexel the reference texel as well
+							}
+						}
+					};
+
+					//uint8_t* imageSumOutData = nullptr; // TODO
+					//memcpy(commonExecuteData.outData+commonExecuteData.oit->getByteOffset(localOutPos,commonExecuteData.outByteStrides),commonExecuteData.inData+readBlockArrayOffset,commonExecuteData.outBlockByteSize);
+				};
+
+				CBasicImageFilterCommon::executePerRegion(commonExecuteData.inImg, sum, commonExecuteData.inRegions.begin(), commonExecuteData.inRegions.end(),clip);
```

Comment:
the only lambda that you should be executing here should be one to decode the pixels into your temporary scratch memory store

Reply:
Upheld, right now you have a huge problem that you do decode,sum,encode inside your per-region lambda... this will backfire for sparse images.

First everything needs to be decoded from all regions, then summed.. ONCE, then encoded to all regions

Reply:
what do you mean in more detail, could you explain? 

> this will backfire for sparse images. 

but if the user choose the mipmap and a layer, only one region will be ever executed (afaik), so why can't I do it in per region lambda?

Reply:
scratch memory is independent, it will always hold one layer with no reference, so if I compute offset to fetch texels correctly from the region being executed, I can (probably, maybe I'm wrong) decode, sum and encode in the lambda as I think of it

Reply:
solved